### PR TITLE
AP_TempratureSensor: add support for analog sensor with polynomial

### PR DESCRIPTION
**Pull request by IamPete1**
This work is sponsored by Spektreworks.

This adds support for a analog voltage temperature sensor. It uses a 4th order polynomial to go from voltage to temperature for use with thermistors.  Because of the 5 polynomial terms and the pin this adds a var pointer param table to AP_TempratureSensor like those in battery and rangefinder.

### CODE REVIEW

File: libraries/AP_TemperatureSensor/AP_TemperatureSensor_Analog.cpp [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "AP_TemperatureSensor_Analog.h"
+
+#if AP_TEMPERATURE_SENSOR_ANALOG
```
Comment (peterbarker) [inline libraries/AP_TemperatureSensor/AP_TemperatureSensor_Analog.cpp:18]:
```suggestion
#if AP_TEMPERATURE_SENSOR_ANALOG_ENABLED
```

File: libraries/AP_TemperatureSensor/AP_TemperatureSensor_Analog.h [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+#include "AP_TemperatureSensor_Backend.h"
+
+#if AP_TEMPERATURE_SENSOR_ANALOG
```
Comment (peterbarker) [inline libraries/AP_TemperatureSensor/AP_TemperatureSensor_Analog.h:19]:
```suggestion
#include "AP_TemperatureSensor_config.h"

#if AP_TEMPERATURE_SENSOR_ANALOG

#include "AP_TemperatureSensor_Backend.h"
```

File: libraries/AP_TemperatureSensor/AP_TemperatureSensor_Analog.cpp [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+extern const AP_HAL::HAL &hal;
+
+const AP_Param::GroupInfo AP_TemperatureSensor_Analog::var_info[] = {
+
+    // @Param: PIN
+    // @DisplayName: Voltage sensing pin
```
Comment (rmackay9) [inline libraries/AP_TemperatureSensor/AP_TemperatureSensor_Analog.cpp:25]: Normally we should have the subsystem in the parameter display name.

File: libraries/AP_TemperatureSensor/AP_TemperatureSensor_Analog.h (hunk 2 above)
Comment (peterbarker) [reply libraries/AP_TemperatureSensor/AP_TemperatureSensor_Analog.h:19]: ... similarly in the .cpp file

**Review by rmackay9 — APPROVED** [review]

**Review by peterbarker — APPROVED** [review]

**Comment by rmackay9** [comment]
It would be nice to get @magicrub's approval if he's tested.

File: libraries/AP_TemperatureSensor/AP_TemperatureSensor.cpp [hunk 4]
```diff
@@ -147,6 +190,11 @@ void AP_TemperatureSensor::init()
             case AP_TemperatureSensor_Params::Type::TSYS03:
                 drivers[instance] = new AP_TemperatureSensor_TSYS03(*this, _state[instance], _params[instance]);
                 break;
+#endif
+#if AP_TEMPERATURE_SENSOR_TSYS03_ENABLED
```
Comment (magicrub) [inline libraries/AP_TemperatureSensor/AP_TemperatureSensor.cpp:194]: Wrong define

File: libraries/AP_TemperatureSensor/AP_TemperatureSensor_Analog.cpp [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+void AP_TemperatureSensor_Analog::init()
+{
+    _analog_source = hal.analogin->channel(_pin);
+}
+
+// Update function called at 5Hz
+void AP_TemperatureSensor_Analog::update()
+{
+    if (!_analog_source->set_pin(_pin)) {
```
Comment (magicrub) [inline libraries/AP_TemperatureSensor/AP_TemperatureSensor_Analog.cpp:76]: Needs Null check

File: libraries/AP_TemperatureSensor/AP_TemperatureSensor_Analog.cpp [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+                                                         AP_TemperatureSensor_Params &params) :
+    AP_TemperatureSensor_Backend(front, state, params)
+{
+    AP_Param::setup_object_defaults(this, var_info);
+    _state.var_info = var_info;
+}
+
+void AP_TemperatureSensor_Analog::init()
+{
+    _analog_source = hal.analogin->channel(_pin);
```
Comment (magicrub) [inline libraries/AP_TemperatureSensor/AP_TemperatureSensor_Analog.cpp:70]: Might want to just skip the init stuff and run it in the update

File: libraries/AP_TemperatureSensor/AP_TemperatureSensor_Analog.cpp [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    if (!_analog_source->set_pin(_pin)) {
+        // Invalid pln
+        return;
+    }
+
+    // Use ratiometric voltage, measured voltage is relative to supply
+    const float voltage = _analog_source->voltage_average_ratiometric();
+
+    // Evaluate polynomial
+    float temp = 0.0;
```
Comment (magicrub) [inline libraries/AP_TemperatureSensor/AP_TemperatureSensor_Analog.cpp:85]: Would be great to have the polynomial equation here in a comment

File: libraries/AP_TemperatureSensor/AP_TemperatureSensor_Analog.cpp (hunk 1 above)
Comment (magicrub) [reply libraries/AP_TemperatureSensor/AP_TemperatureSensor_Analog.cpp:18]: I second this. Needs enabled in the define name

**Review by magicrub — CHANGES_REQUESTED** [review]

**Comment by magicrub** [comment]
@rmackay9 I don't have a sensor to test with.

**Review by peterbarker — CHANGES_REQUESTED** [review]
... my original review should have been request-changes...

File: libraries/AP_TemperatureSensor/AP_TemperatureSensor.cpp (hunk 4 above)
Comment (IamPete1) [reply libraries/AP_TemperatureSensor/AP_TemperatureSensor.cpp]: Yep, fixed

File: libraries/AP_TemperatureSensor/AP_TemperatureSensor_Analog.cpp (hunk 5 above)
Comment (IamPete1) [reply libraries/AP_TemperatureSensor/AP_TemperatureSensor_Analog.cpp]: added one.

File: libraries/AP_TemperatureSensor/AP_TemperatureSensor_Analog.cpp (hunk 7 above)
Comment (IamPete1) [reply libraries/AP_TemperatureSensor/AP_TemperatureSensor_Analog.cpp:85]: added

File: libraries/AP_TemperatureSensor/AP_TemperatureSensor_Analog.h (hunk 2 above)
Comment (IamPete1) [reply libraries/AP_TemperatureSensor/AP_TemperatureSensor_Analog.h]: Done.

File: libraries/AP_TemperatureSensor/AP_TemperatureSensor_Analog.cpp (hunk 1 above)
Comment (IamPete1) [reply libraries/AP_TemperatureSensor/AP_TemperatureSensor_Analog.cpp]: Done

**Review by magicrub — APPROVED** [review]
LGTM.. although I'm not in favor of the name, it feels misleading. It's not just "analog", it's specific to a type of signal that is output by a specific sensor. It's not just a plain old analog voltage like we've used in rangefinder and battery stuff in the past.

**Comment by IamPete1** [comment]
@magicrub I'm open to better suggestions.

The analog range finder has a offset and a scale factor, that is the first two params of our polynomial. It also has a param for a fit type, "LINEAR", "INVERTED" and "HYPERBOLA". I don't think this is so different to that.

**Review by tridge — APPROVED** [review]

**Pull request merged**